### ISSUE TITLE
[Go target] Generated lexer uses "l" for some functions, "p" for other functions.

### ISSUE THREAD

**Issue by kaby76** [antlr/antlr4#4306]
This is in regards to writing a "target agnostic grammar", where we try to reuse a grammar even though actions ("semantic predicates") in a grammar are always target specific.

The [javascript](https://github.com/antlr/grammars-v4/tree/master/javascript/javascript) grammar contains action code for tokenization. For example, [here](https://github.com/antlr/grammars-v4/blob/6590fdd07b9fa1b4a66025c8315ee2c93bbcbe7d/javascript/javascript/JavaScriptLexer.g4#L48). Notice that the `"this."` prefix is platform specific syntax for a qualified method call in Java, as well as CSharp, Dart, JavaScript (modern), and TypeScript. For the other targets, the grammar needs t be `sed -i` of the .g4's to get them to work (and they do--"target agnostic format" really works). For PHP, it needs to be changed to `"self::'"`. For Python3, `"self."`. For Cpp, `"this->"`. For Go, it's much, much harder.

The reason it is harder is because the "methods" are implemented using functions that pass a pointer to the type.

```
func (l *JavaScriptLexer) OpenBrace_Action(localctx antlr.RuleContext, actionIndex int) {
	switch actionIndex {
	case 0:
		l.ProcessOpenBrace()

	default:
		panic("No registered action for: " + fmt.Sprint(actionIndex))
	}
}
```

That would be fine--if I knew that `"this."` should be rewritten always to `"l."` because the parameter in the function is named "l". But it is not always the case. For [this action in the lexer grammar](https://github.com/antlr/grammars-v4/blob/6590fdd07b9fa1b4a66025c8315ee2c93bbcbe7d/javascript/javascript/JavaScriptLexer.g4#L39), the parameter is named "p".

```
func (p *JavaScriptLexer) HashBangLine_Sempred(localctx antlr.RuleContext, predIndex int) bool {
	switch predIndex {
	case 0:
		return l.IsStartOfFile()

	default:
		panic("No predicate with index: " + fmt.Sprint(predIndex))
	}
}
```

This is rather annoying because rather than use a parameter named "this" for all these generated functions in the parser and lexer, I need to modify the grammar to use the name that is used in the templates. I have to look at the templates for Go, and see where "p" vs "l" are used, then change the action code to use that name.

Could we change the Go templates just always use "this" as a parameter name?

**Comment by jimidle** [comment]
I can make it consistent (probably) but this is not a pattern for go. I
imagine that the original template copied code from the parse gen to the
lexer and forgot to change the receiver. It is idiomatic to use the same
receiver name for all funcs. It would normally be p for parser and l for
lexer.

I’ll take a look

**Comment by kaby76** [comment]
This would be a "nice to have", but not critical. I can use Trash to extract plain actions vs semantic predicate actions, modify those actions with "l." or "p." accordingly, then reconstruct the grammar. The build tools can handle this. I just don't quite get why it's not a consistent name like "self" in the generated code, instead of all this "l" or "p".

**Comment by jimidle** [comment]
self or this are different concepts really. But we should be using consistent receiver variable names.

**Comment by kaby76** [comment]
Actually, sort of is one, which I accidentally stumbled upon it with a typo I wrote. https://github.com/antlr/antlr4/issues/3508 `$parser`. (It's still pretty much an undocumented feature, except for [this added last year](https://github.com/antlr/antlr4/blob/b7b8135ff9f409c19fc24ee962600c56ec3467db/doc/python-target.md?plain=1#L123) in the python target documentation.) `$parser` "works", but there are two flaws. One is that it's only good for the parser, i.e., there's no `$lexer` object reference. The second problem is that for Cpp, the syntax for object dereference uses the `->` operator, not `.`. So, for it to work in the Cpp target, the action must be `$parser->`, not `$parser.`, which is fine for all the other targets.

**Comment by ericvergnaud** [comment]
The cpp version could be a reference rather than a pointer, that would avoid the -> vs . issue.

**Comment by kaby76** [comment]
> The cpp version could be a reference rather than a pointer, that would avoid the -> vs . issue.

Yes, instead of `$parser->foobar()`, use `$parser.foobar()`, and in the Antlr tool, generate code as `(*this).foobar()`. That would work for at least `$parser`. Still missing it for the lexer. Either need to add `$lexer` or better, `$this` and always use the `.` operator afterwards.

**Comment by kaby76** [comment]
Yes, we could use `$parser` and `$lexer` (or `$this`, which I would need to think about what you said), then I could pitch in most cases the transformGrammar.py used in grammars-v4.

I'll still need the transformGrammar.py hack because there are a couple of other issues where it corrects other problems with codegen. In fortran/fortran90, I have to sed [this line](https://github.com/antlr/grammars-v4/blob/8d58426f64b98637d55f6e49a51544bac5106b5a/fortran/fortran90/Fortran90Lexer.g4#L5) to turn it into `@header {#include "Fortran90LexerBase.h"}` for the Cpp target, and `@header {require "Fortran90LexerBase.php";}` for the PHP target. Honestly, we go to the trouble to declare in a grammar file [options { superClass = Fortran90LexerBase; }](https://github.com/antlr/grammars-v4/blob/8d58426f64b98637d55f6e49a51544bac5106b5a/fortran/fortran90/Fortran90Lexer.g4#L3), and the tool doesn't generate the proper `#include's` or 'import's` to work. Why is that?

**Comment by ericvergnaud** [comment]
Well I guess whilst it's obviously good practice to name the abstract parser header from its name, it's definitely not something we can enforce. We're not generating or providing that file are we ?
I would personally be ok to enforce the naming, but that might break a number of working grammars, notably in the frequent case where developers place the generated files in a dedicated folder...
I'm personally supportive of a macro system, where grammar code would be conditionally included based on the active target

#if PHP
require "Fortran90LexerBase.php"
#elif CPP
#include "Fortran90LexerBase.h"
#endif

IIRC you hate the idea (or maybe that's Ivan?), but tbh I haven't seen a better one :-(

**Comment by kaby76** [comment]
This would be okay. I think my concern was lexing the target code between the `#if` and `else`. The target code itself may have Cpp preprocessor directives:
```
grammar Foobar;

#if Cpp
// Cpp code   vvv.
#if defined(XXX)
#include "xxx.h"
#else
#include "other.h"
#endif  
// Cpp code ^^^
#endif  
```
It just makes the lexing more difficult, e.g., is the `#else` part of the Cpp target code, or `@member` code for all other targets? `@member` uses an action block, so it only has to count and match open and closing braces outside of string literals within the action blocks.

**Comment by ericvergnaud** [comment]
We could certainly pick a different pattern but tbh if developers are clever enough to introduce such complexity then I expect they'd also be able to fix them.

**Comment by ericvergnaud** [comment]
I believe we need this multi target capability in more than 1 place, but the curly brace encloser is present with all of themHow about @member(CPP) { … ***@***.***(CPP) { … } And so forth ?Envoyé de mon iPhoneLe 8 juin 2023 à 16:51, Ken Domino ***@***.***> a écrit :﻿
This would be okay. I think my concern was lexing the target code between the #if and else. The target code itself may have Cpp preprocessor directives:
grammar Foobar;

#if Cpp
// Cpp code   vvv.
#if defined(XXX)
#include "xxx.h"
#else
#include "other.h"
#endif  
// Cpp code ^^^
#endif  

It just makes the lexing more difficult, e.g., is the #else part of the Cpp target code, or @member code for all other targets? @member uses an action block, so it only has to count and match open and closing braces outside of string literals within the action blocks.

**Comment by kaby76** [comment]
That's cool. Works for me!

**Comment by KvanTTT** [comment]
> IIRC you hate the idea (or maybe that's Ivan?), but tbh I haven't seen a better one :-(

Yes, I still think the idea with preprocessor directives is not good. I have had experience with Objective-C language processing and encountered a lot of problems with preprocessor directives. Probably you meant this topic: [Support predicates](https://github.com/antlr/antlr4-lab/discussions/73#discussioncomment-5123652). I suggest using something like this (it was suggested by @udif in [Unified Actions Language](https://github.com/antlr/antlr4/discussions/4067#discussioncomment-4648531)):

```antlr
stat
    : { java5() }? 'goto' ID ';'

...

@parser::members::java {
  boolean java5() {
    ...
  }
}
@parser::members::cpp {
  bool java5() {
    ...
  }
}
```

We can use universal function calls and target-specific function declarations.

BTW, currently I'm working on [Multiplatform feature](https://kotlinlang.org/docs/multiplatform.html) in Kotlin and it looks like the idea above (in Kotlin there are `expect` and `actual` functions).

**Comment by kaby76** [comment]
> We can use universal function calls and target-specific function declarations.

Exactly! Thank you @KvanTTT. Anltr already has `$parser`, _which somehow someone snuck into the Antlr tool_. And, we also have syntax for accessing attributes in actions. But, now we have an impasse where can't agree on something like `$parser` for lexers, and the translation of the operator that follows the reference into the target-specific syntax in order to make it a little more bearable over in grammars-v4. We don't need to devise a general-purpose language for actions, just a common way to access code in the target. That's a plane of features that I can use.

**Comment by ericvergnaud** [comment]
Mmm… the bad news is that whilst this may work for Java which has an implicit this, it won’t for other targets, notably JS, TS and Python…Envoyé de mon iPhoneLe 10 juin 2023 à 17:06, Ken Domino ***@***.***> a écrit :﻿

We can use universal function calls and target-specific function declarations.

Exactly! Thank you @KvanTTT. Anltr already has $parser, which somehow someone snuck into the Antlr tool. And, we also have syntax for accessing attributes in actions. But, now we have an impasse where can't agree on something like $parser for lexers, and the translation of the operator that follows the reference into the target-specific syntax in order to make it a little more bearable over in grammars-v4. We don't need to devise a general-purpose language for actions, just a common way to access code in the target. That's a plane of features that I can use.

**Comment by ericvergnaud** [comment]
(Which is why I had to introduce $parser in the first place)Envoyé de mon iPhoneLe 11 juin 2023 à 12:50, Wanadoo ***@***.***> a écrit :﻿Mmm… the bad news is that whilst this may work for Java which has an implicit this, it won’t for other targets, notably JS, TS and Python…Envoyé de mon iPhoneLe 10 juin 2023 à 17:06, Ken Domino ***@***.***> a écrit :﻿

We can use universal function calls and target-specific function declarations.

Exactly! Thank you @KvanTTT. Anltr already has $parser, which somehow someone snuck into the Antlr tool. And, we also have syntax for accessing attributes in actions. But, now we have an impasse where can't agree on something like $parser for lexers, and the translation of the operator that follows the reference into the target-specific syntax in order to make it a little more bearable over in grammars-v4. We don't need to devise a general-purpose language for actions, just a common way to access code in the target. That's a plane of features that I can use.

**Comment by kaby76** [comment]
> Mmm… the bad news is that whilst this may work for Java which has an implicit this, it won’t for other targets, notably JS, TS and Python.

Not sure what you are looking at, but "this." is perfectly valid for JavaScript. Here's the constructor code that's generated for the lexer.
```
    constructor(input) {
        super(input)
        this._interp = new antlr4.atn.LexerATNSimulator(this, atn, decisionsToDFA, new antlr4.atn.PredictionContextCache());
    }
```

[Generated-JavaScript-0.zip](https://github.com/antlr/antlr4/files/11714429/Generated-JavaScript-0.zip)

I can prove it for TypeScript and Python, too, if you need to see a functioning, complete example.

**Comment by kaby76** [comment]
Yes, you are right for Python3. There is no "this." But, the code passes a "self" in each of the wrapper functions. So, my "transformGrammar.py" hack rewrites the "this." to "self.". The generated code also works fine.

[Generated-Python3-0.zip](https://github.com/antlr/antlr4/files/11714442/Generated-Python3-0.zip)

**Comment by KvanTTT** [comment]
> Yes, you are right for Python3. There is no "this." But, the code passes a "self" in each of the wrapper functions. So, my "transformGrammar.py" hack rewrites the "this." to "self.". The generated code also works fine.

To be precise in Python use can use any word as receiver, not only `self`:

```py
class C:
    def f(foo):
        print(foo.x)

c = C()
c.x = 42
c.f() // prints 42
```

**Comment by KvanTTT** [comment]
> Mmm… the bad news is that whilst this may work for Java which has an implicit this, it won’t for other targets, notably JS, TS and Python…

It will work work other targets if consider first parameter as receiver (actually on IR level all class members represented as members with first `this` parameter).

**Comment by ericvergnaud** [comment]
Let's move this to a discussion ?

**Comment by KvanTTT** [comment]
Yes but I suggest continuing existing ones: [Unified Actions Language](https://github.com/antlr/antlr4/discussions/4067) or [Support predicates](https://github.com/antlr/antlr4-lab/discussions/73).

**Comment by kaby76** [comment]
I'm moving this to the discussions as suggested. But I recommend looking at the hacks that I have to employ in order to get a grammar to work across 8 different targets [here](https://github.com/antlr/grammars-v4/tree/354f608994c683c5d3530fc1209394458cb34020/fortran/fortran90).

**Comment by kaby76** [comment]
(I can't convert the issue to a discussion.)
